mantle/kola/tests/rhcos/upgrade.go

```diff
@@ -165,3 +188,190 @@ func rhcosUpgradeBasic(c cluster.TestCluster) {
+		})
+	default:
+		c.Fatal("Platform unsupported")
+	}
+
+}
+
+// Downloads the latest RHCOS from an OCP stream and decompresses it.
+// Returns the path to the decompressed file
+func downloadLatestReleasedRHCOS(target string) (string, error) {
+	buildID := kola.CosaBuild.Meta.BuildID
+	ocpVersion := strings.Split(buildID, ".")[0]
+	ocpVersionF := fmt.Sprintf("%s.%s", ocpVersion[:1], ocpVersion[1:])
+	channel := "fast-" + ocpVersionF
+
+	type Release struct {
+		Version string `json:"version"`
+		Payload string `json:"payload"`
+	}
+
+	type Graph struct {
+		Nodes []Release `json:"nodes"`
+		Edges [][]int   `json:"edges"`
+	}
+
+	type MachineOS struct {
+		Version     string `json:"Version"`
+		DisplayName string `json:"DisplayName"`
+	}
+
+	type DisplayVersions struct {
+		MachineOS MachineOS `json:"machine-os"`
+	}
+
+	type OcpRelease struct {
+		DisplayVersions DisplayVersions `json:"displayVersions"`
+	}
+
+	graph := &Graph{}
+	graphUrl := fmt.Sprintf("https://api.openshift.com/api/upgrades_info/v1/graph?channel=%s", channel)
+	util.GetJson(graphUrl, &graph)
+
+	// no-op on unreleased OCP versions
+	if len(graph.Nodes) == 0 {
+		return "", nil
+	}
+
+	// Find the latest OCP release by looking at the edges and comparing it to
+	// the nodes. Edges define updates as [from release index, to release index]
+	// so the node that doesn't show up in a from release index is the latest.
+	fromEdge := []int{}
+	releaseIndex := []int{}
+	for i := 0; i < len(graph.Nodes); i++ {
+		releaseIndex = append(releaseIndex, i)
+	}
+
+	for _, v := range graph.Edges {
+		fromEdge = append(fromEdge, v[0])
+	}
+
+	unique := func(intSlice []int) []int {
+		keys := make(map[int]bool)
+		list := []int{}
+		for _, entry := range intSlice {
+			if _, value := keys[entry]; !value {
+				keys[entry] = true
+				list = append(list, entry)
+			}
+		}
+		return list
+	}(fromEdge)
+
+	difference := func(a, b []int) (diff []int) {
+		m := make(map[int]bool)
+		for _, item := range b {
+			m[item] = true
+		}
+
+		for _, item := range a {
+			if _, ok := m[item]; !ok {
+				diff = append(diff, item)
+			}
+		}
+		return
+	}(releaseIndex, unique)
+
+	// The origin-clients package in Fedora doesn't `oc adm release info`
+	// ability.
+	ocUrl := "https://mirror.openshift.com/pub/openshift-v4/clients/ocp/latest/openshift-client-linux.tar.gz"
+	cmdString := fmt.Sprintf("curl -s %s | sudo tar -zxvf - -C /usr/bin", ocUrl)
+	if err := exec.Command("bash", "-c", cmdString).Run(); err != nil {
+		return "", err
+	}
+
+	var ocpRelease *OcpRelease
+	latestOcpPayload := graph.Nodes[difference[0]].Payload
+	cmd := exec.Command("oc", "adm", "release", "info", latestOcpPayload, "-o", "json")
+	output, err := cmd.Output()
+	if err != nil {
+		return "", err
+	}
+	json.Unmarshal(output, &ocpRelease)
+
+	var latestOcpRhcosBuild *cosa.Build
+	rhcosVersion := ocpRelease.DisplayVersions.MachineOS.Version
+	latestBaseUrl := fmt.Sprintf("https://art-rhcos-ci.s3.amazonaws.com/releases/rhcos-%s/%s/%s",
```

Comment:
Inside coreos-assembler, we already have code which knows how to parse the stream metadata, which is the intended "official" API that we should use here too I think going forward.

This is documented here https://github.com/openshift/installer/blob/master/docs/user/overview.md#coreos-bootimages

(Also tangentially related to this...we could create e.g. openshift/coreos-assembler which would be a layered docker image where we'd stick things like `oc` and `openshift-install` by default too...)

Reply:
See also this code https://github.com/coreos/coreos-assembler/blob/main/mantle/rhcos/metadata.go which parses the git repo - it only supports a `main` rhcos stream right now but we could easily teach it to find data from the `release-$x` branches.